### PR TITLE
[33] Implement edit job API @PUT

### PR DESCRIPTION
**Pull request by badnameok**
## What did you do

- Implemented edit job feature
- also includes both successful and forbidden test units
- fixed a few bugs/unused code from post-job as well

## Demo

- login into service as caster
```
register
{
  "email": "ayame@hololive.com",
  "password": "ayame123",
  "firstName": "Ayame",
  "lastName": "Nakiri",
  "prefix": "Mrs.",
  "nationality": "Japanese",
  "ssn": "1212312121",
  "gender": "FEMALE",
  "phoneNumber": "0123456789",
  "idCardImageUrl": "hololive.com"
}

login
{
  "email": "yagoo@hololive.com",
  "password": "yagoo123"
}
```
- post job
- put job with the following
```
sample edit
{
  "title": "NEW UPDATE",
  "description": "IT WORKS",
  "status": "OPEN",
  "role": "WORKS",
  "minAge": 1,
  "maxAge": 5,
  "gender": "ANY",
  "actorCount": 1,
  "wage": 555555,
  "applicationDeadline": "2024-02-11T17:42:50.556Z",
  "shooting": [
    {
      "shootingLocation": "string",
      "startDate": "2024-02-11T17:42:50.556Z",
      "endDate": "2024-02-11T17:42:50.556Z",
      "startTime": "2024-02-11T17:42:50.556Z",
      "endTime": "2024-02-11T17:42:50.556Z"
    }
  ]
}
```
- pnpm studio

### CODE REVIEW

File: apps/api/src/modules/job/job.service.spec.ts [hunk 1]
```diff
@@ -355,149 +405,87 @@ describe('JobService', () => {
...
 
         const newId = await service.createJob(MOCK_JOB, MOCK_CASTING_ID)
 
         expect(repository.createJob).toBeCalledWith(MOCK_JOB, MOCK_CASTING_ID)
       })
+
+      it('should be bad request due to date conflict', async () => {
+        const MOCK_JOB = createValidJob()
+
+        const result = mock('job').get(1)[0]
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:422]:
```suggestion
        const result = mock('job').get()
```
https://www.notion.so/pkhing/Testing-e6908a78c4f94c2f9c8e3510130332d4?pvs=4#7393d7aef39b4186a3aed616f64259a1

File: packages/dtos/src/job.ts [hunk 2]
```diff
@@ -235,6 +179,10 @@ export class CreateJobDto {
   shooting: ShootingDto[]
 }
 
+export class EditJobDto extends CreateJobDto implements EditJobType {
```
Comment (palmcm) [inline packages/dtos/src/job.ts:182]: CreateJobDto should be implemented the type instead, If you put it here there are no need to use it. the implements type is use for type checking with the database only

File: apps/api/src/modules/job/job.service.spec.ts [hunk 3]
```diff
@@ -355,149 +405,87 @@ describe('JobService', () => {
...
+
+        jest.spyOn(repository, 'createJob').mockResolvedValue(result.jobId)
+
+        await expect(
+          service.createJob(MOCK_JOB, MOCK_CASTING_ID),
+        ).rejects.toThrow(BadRequestException)
+
+        // expect repository.createJob to not be called
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:430]: remove this comment?

File: apps/api/src/modules/job/job.service.spec.ts [hunk 4]
```diff
@@ -355,149 +405,87 @@ describe('JobService', () => {
...
-        }
-        if (MOCK_JOB.wage < 1) {
-          MOCK_JOB.wage = 1
-        }
+    describe('normal behavior', () => {
+      it('should update the job successfully', async () => {
+        const MOCK_UPDATED_TITLE = 'updated title'
+        const MOCK_JOB = createValidJob()
 
         const result = mock('job').get(1)[0]
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:444]:
```suggestion
        const result = mock('job').get()
```
https://www.notion.so/pkhing/Testing-e6908a78c4f94c2f9c8e3510130332d4?pvs=4#7393d7aef39b4186a3aed616f64259a1

File: apps/api/src/modules/job/job.service.spec.ts [hunk 5]
```diff
@@ -355,149 +405,87 @@ describe('JobService', () => {
...
-        console.log(result)
-
         jest.spyOn(repository, 'createJob').mockResolvedValue(result.jobId)
+        jest.spyOn(repository, 'updateJob').mockResolvedValue(result.jobId)
 
         const newId = await service.createJob(MOCK_JOB, MOCK_CASTING_ID)
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:448]: service.createJob will do nothing except return `result.jobId` since you already mock id

**Review by PKhing — COMMENTED** [review]
Please fix GetJobDto since it extends EditJobDto

File: apps/api/src/modules/job/job.service.ts [hunk 6]
```diff
@@ -235,8 +239,16 @@ export class JobService {
     return jobDetail
   }
 
-  update(id: number, updateJobDto: EditJobDto) {
-    return `This action updates a #${id} job`
+  async update(id: number, updateJobDto: EditJobDto, userId: number) {
+    const job = await this.repository.getJobById(id)
+    if (job.castingId !== userId) {
```
Comment (palmcm) [inline apps/api/src/modules/job/job.service.ts:244]: If there is no job with that ID it should throw NotFound

**Review by palmcm — CHANGES_REQUESTED** [review]

**Review by PKhing — CHANGES_REQUESTED** [review]
image

**Comment by PKhing** [comment]
This branch base is not beta. Please Rebase

File: apps/api/src/modules/job/job.service.spec.ts (hunk 5 above)
Comment (palmcm) [reply apps/api/src/modules/job/job.service.spec.ts:448]: You don't need to mock and run createJob in updateJob test

**Comment by PKhing** [comment]
And please merge this branch into dev (read git convention)
https://www.notion.so/pkhing/Git-Convention-61787f42507a49298a456f6f2969ac6e?pvs=4#c0290a30b2f64ccfab1013fca3b1c80e

File: apps/api/src/modules/job/job.service.ts [hunk 7]
```diff
@@ -235,8 +239,16 @@ export class JobService {
...
+      throw new ForbiddenException(
+        "You don't have permission to update this job",
+      )
+    }
+
+    this.validateJobDto(updateJobDto)
+    return this.repository.updateJob(id, updateJobDto, userId)
```
Comment (palmcm) [inline apps/api/src/modules/job/job.service.ts:251]: this service (createJob too) should return in 
```
{
   jobId: number
}
```
or read swagger or read above controller for which type to return

**Review by palmcm — CHANGES_REQUESTED** [review]

**Comment by PKhing** [comment]
image

**Comment by PKhing** [comment]
image

**Comment by PKhing** [comment]
image

**Comment by PKhing** [comment]
image

**Comment by PKhing** [comment]
image

**Comment by PKhing** [comment]
image

**Comment by PKhing** [comment]
image
write test for these case

**Comment by badnameok** [comment]
not done yet still missing most of requested test unit, pushed to test

File: apps/api/src/modules/job/job.repository.ts [hunk 8]
```diff
@@ -1,18 +1,18 @@
...
-import { ForbiddenException, Injectable } from '@nestjs/common'
+import { Injectable } from '@nestjs/common'
 import { PrismaService } from 'src/database/prisma.service'
 
 @Injectable()
 export class JobRepository {
   constructor(private prisma: PrismaService) {}
 
-  async createJob(createJobDto: CreateJobDto, userId: number) {
+  async createJob(createJobDto: EditJobDto, userId: number) {
```
Comment (PKhing) [inline apps/api/src/modules/job/job.repository.ts:15]: ?????

**Review by PKhing — CHANGES_REQUESTED** [review]

**Review by PKhing — CHANGES_REQUESTED** [review]
image

**Review by PKhing — APPROVED** [review]

**Review by palmcm — APPROVED** [review]
No test NotFound case. please add.

**Pull request merged**